File: crates/e2m2e-forces/src/forces/srp.rs
//! SolarRadiationPressure + ConicalShadowModel Rust 移植（仅 `spice` feature）。
//!
//! 1:1 移植自 Python ``srp.py`` + ``shadow.py``：
//! - SRP 用 Montenbruck & Gill eq. 3.75 cannonball 模型
//! - 阴影用 M&G §3.4.2 圆锥几何（视角径 a/b、角距 c、四分支判定 + 精确圆面
//!   重叠面积），多遮挡体按 GMAT GMT-6543 合成

use e2m2e_spice::spice_ffi::{spkezr, SpiceFfiError};

/// SRP 常量（与 Python srp.py 一致）。
const P_SRP_1AU: f64 = 4.56e-6; // N/m²
const AU_KM: f64 = 1.49597870700e8; // km
const KM_TO_M: f64 = 1000.0;

/// 天体半径表（km，与 Python shadow._BODY_RADII_KM 一致）。
pub fn body_radius(body: &str) -> Option<f64> {
    match body {
        "SUN" => Some(695700.0),
        "EARTH" => Some(6378.1363),
        "MOON" => Some(1737.4),
        "MARS" => Some(3396.19),
        "JUPITER" => Some(71492.0),
        "VENUS" => Some(6051.8),
        "MERCURY" => Some(2440.53),
        "SATURN" => Some(60268.0),
        "URANUS" => Some(25559.0),
        "NEPTUNE" => Some(24764.0),
        _ => None,
    }
}

/// 单遮挡体的光照份额（Montenbruck & Gill §3.4.2 圆锥模型）。
///
/// 返回 [0, 1]，1 = 全光照，0 = 全阴影。与 Python shadow._body_flux_factor 一致。
fn body_flux_factor(
    sc: &[f64; 3],
    body: &[f64; 3],
    sun: &[f64; 3],
    body_radius: f64,
    sun_radius: f64,
) -> f64 {
    // sc_to_body, sc_to_sun
    let sc_to_body = [body[0] - sc[0], body[1] - sc[1], body[2] - sc[2]];
    let sc_to_sun = [sun[0] - sc[0], sun[1] - sc[1], sun[2] - sc[2]];
    let d_body = norm(&sc_to_body);
    let d_sun = norm(&sc_to_sun);

    // GMAT 守卫
    if sun_radius >= d_sun {
        return 1.0;
    }
    if body_radius >= d_body {
        return 0.0;
    }

    let a = (sun_radius / d_sun).asin(); // 太阳视角径
    let b = (body_radius / d_body).asin(); // 遮挡体视角径
    let cos_c = dot(&sc_to_body, &sc_to_sun) / (d_body * d_sun);
    let cos_c = cos_c.clamp(-1.0, 1.0);
    let c = cos_c.acos();

    if a + b <= c {
        return 1.0; // 全光照
    }
    if c <= (a - b).abs() {
        if b >= a {
            return 0.0; // 本影
        }
        return 1.0 - (b / a).powi(2); // 环形食
    }
    // 半影：M&G eq. 3.92-3.94
    let a2 = a * a;
    let b2 = b * b;
    let x = (c * c + a2 - b2) / (2.0 * c);
    let y_sq = (a2 - x * x).max(0.0);
    let y = y_sq.sqrt();
    let area = a2 * (x / a).acos() + b2 * ((c - x) / b).acos() - c * y;
    1.0 - area / (std::f64::consts::PI * a2)
}

/// 多遮挡体合成（GMAT GMT-6543，与 Python _combine_body_fluxes 一致）。
///
/// `factors[i]` 是第 i 个遮挡体的光照份额（0..1）；
/// `angular_radii[i]` 是遮挡体视角径；`directions[i]` 是 sc→遮挡体 单位向量。
fn combine_body_fluxes(factors: &[f64], angular_radii: &[f64], directions: &[[f64; 3]]) -> f64 {
    let n = factors.len();
    if n == 0 {
        return 1.0;
    }
    if n == 1 {
        return factors[0];
    }
    // 任一本影（factor=0）→ 0
    if factors.contains(&0.0) {
        return 0.0;
    }
    if n == 2 {
        // 两体：检查日盘上重叠
        let a_i = angular_radii[0];
        let a_j = angular_radii[1];
        let cos_c_ij = dot(&directions[0], &directions[1]); // 单位向量点积 = cos
        let c_ij = cos_c_ij.acos();
        if a_i + a_j < c_ij {
            // 日盘上不重叠
            return factors[0] + factors[1] - 1.0;
        }
        // 重叠
        return factors[0].min(factors[1]);
    }
    // 3+ 体：取 min
    let mut min_f = factors[0];
    for &f in &factors[1..] {
        if f < min_f {
            min_f = f;
        }
    }
    min_f
}

/// ConicalShadowModel 完整光照份额（系统感知）。
///
/// 与 Python ConicalShadowModel.flux_factor 一致。返回 [0, 1]。
pub fn flux_factor(
    et: f64,
    sc_pos: &[f64; 3],
    shadow_bodies: &[String],
    observer: &str,
) -> Result<f64, SpiceFfiError> {
    if shadow_bodies.is_empty() {
        return Ok(1.0);
    }
    // 查太阳 + 各遮挡体相对 observer 的 J2000 位置
    let (sun_state, _) = spkezr("SUN", et, "J2000", "NONE", observer)?;
    let sun_pos = [sun_state[0], sun_state[1], sun_state[2]];
    let sun_radius = body_radius("SUN").unwrap();

    let mut factors: Vec<f64> = Vec::with_capacity(shadow_bodies.len());
    let mut angular_radii: Vec<f64> = Vec::with_capacity(shadow_bodies.len());
    let mut directions: Vec<[f64; 3]> = Vec::with_capacity(shadow_bodies.len());

    for body in shadow_bodies {
        let body_radius = match body_radius(body) {
            Some(r) => r,
            None => continue,
        };
        let (body_state, _) = spkezr(body, et, "J2000", "NONE", observer)?;
        let body_pos = [body_state[0], body_state[1], body_state[2]];
        let f = body_flux_factor(sc_pos, &body_pos, &sun_pos, body_radius, sun_radius);
        factors.push(f);

        let sc_to_body = [
            body_pos[0] - sc_pos[0],
            body_pos[1] - sc_pos[1],
            body_pos[2] - sc_pos[2],
        ];
        let d = norm(&sc_to_body);
        angular_radii.push((body_radius / d).min(1.0).asin());
        directions.push([sc_to_body[0] / d, sc_to_body[1] / d, sc_to_body[2] / d]);
    }
    Ok(combine_body_fluxes(&factors, &angular_radii, &directions))
}

/// SRP 加速度（含阴影）。
///
/// 与 Python SolarRadiationPressure.compute_acceleration 一致。
/// 返回 [ax, ay, az] km/s²，在 observer 系（通常 J2000 地心）下。
pub fn srp_acceleration(
    et: f64,
    sc_pos: &[f64; 3],
    area: f64,
    mass: f64,
    cr: f64,
    shadow_bodies: &[String],
    observer: &str,
) -> Result<[f64; 3], SpiceFfiError> {
    let (sun_state, _) = spkezr("SUN", et, "J2000", "NONE", observer)?;
    let sun_pos = [sun_state[0], sun_state[1], sun_state[2]];

    // sun → sc 向量
    let sun_to_sc = [
        sc_pos[0] - sun_pos[0],
        sc_pos[1] - sun_pos[1],
        sc_pos[2] - sun_pos[2],
    ];
    let r = norm(&sun_to_sc);
    if r == 0.0 {
        return Ok([0.0, 0.0, 0.0]);
    }

    let flux = if shadow_bodies.is_empty() {
        1.0
    } else {
        flux_factor(et, sc_pos, shadow_bodies, observer)?
    };

    // SRP 加速度公式（与 Python _compute_srp_acceleration 一致）
    let pressure = P_SRP_1AU * (AU_KM / r).powi(2); // N/m²
    let mag_si = flux * pressure * cr * area / mass; // m/s²
    let mag_km = mag_si / KM_TO_M; // km/s²
    Ok([
        mag_km * sun_to_sc[0] / r,
        mag_km * sun_to_sc[1] / r,
        mag_km * sun_to_sc[2] / r,
    ])
}

fn norm(v: &[f64; 3]) -> f64 {
    (v[0] * v[0] + v[1] * v[1] + v[2] * v[2]).sqrt()
}

fn dot(a: &[f64; 3], b: &[f64; 3]) -> f64 {
    a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flux_no_shadow() {
        // 没有遮挡体，flux = 1
        let f = flux_factor(0.0, &[100000.0, 0.0, 0.0], &[], "EARTH").unwrap();
        assert_eq!(f, 1.0);
    }

    #[test]
    fn srp_basic_finite() {
        // 跑通即可（不加载内核会失败，跳过具体断言）
        let result = srp_acceleration(0.0, &[100000.0, 0.0, 0.0], 10.0, 1000.0, 1.0, &[], "EARTH");
        // 不加载内核会失败，但函数应该正确处理
        assert!(result.is_ok() || result.is_err());
    }
}

File: e2m2e/algorithm/forces/srp.py
"""太阳光压力模型（cannonball / 基础 Cr 系数）。

实现 Montenbruck & Gill eq. 3.75 的 cannonball SRP 模型：

    a = flux_factor · P · (1 AU / r)² · Cr · A / m · û

其中 ``û`` 为 Sun→SC 单位向量（指向远离太阳），``P = 4.56e-6 N/m²`` 为 1 AU
处的太阳光压常数。``flux_factor ∈ [0, 1]`` 由阴影模型给出（全光照=1，本影=0）。

加速度计算全部由 Rust 编译路径承载（``("srp", ...)`` 力元组，
``crates/e2m2e-forces/src/forces/srp.rs``），Python 侧不保留参考实现。

References:
    - Montenbruck & Gill, *Satellite Orbits*, eq. 3.75
    - GMAT R2026a ``SolarRadiationPressure`` (Spherical 模型)
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .physical_model import PhysicalModel

if TYPE_CHECKING:
    from .shadow import ConicalShadowModel


class SolarRadiationPressure(PhysicalModel):
    """基础 Cr 系数太阳光压模型（cannonball）。

    Args:
        area: 航天器迎风截面积，单位 m²。
        mass: 航天器质量，单位 kg。
        cr: 辐射反射系数（1=全吸收，2=全反射），默认 1.5。
        shadow: 阴影模型（注入）。``None`` 表示全光照（flux_factor 恒为 1）。
    """

    def __init__(
        self,
        area: float,
        mass: float,
        cr: float = 1.5,
        shadow: ConicalShadowModel | None = None,
    ) -> None:
        self._area = float(area)
        self._mass = float(mass)
        self._cr = float(cr)
        self._shadow = shadow
        if self._area <= 0:
            raise ValueError("area must be positive")
        if self._mass <= 0:
            raise ValueError("mass must be positive")

    @property
    def area(self) -> float:
        """截面积（m²）。"""
        return self._area

    @property
    def mass(self) -> float:
        """质量（kg）。"""
        return self._mass

    @property
    def cr(self) -> float:
        """辐射反射系数 Cr。"""
        return self._cr

    @property
    def shadow(self) -> ConicalShadowModel | None:
        """注入的阴影模型，``None`` 表示全光照。"""
        return self._shadow

    def to_rust_spec(self, system) -> tuple | None:
        """序列化为 ``("srp", area, mass, cr, shadow_bodies)``。"""
        shadow_bodies = list(self._shadow.bodies) if self._shadow is not None else []
        return ("srp", self._area, self._mass, self._cr, shadow_bodies)


class VariableMassSolarRadiationPressure(PhysicalModel):
    """质量由增广状态提供的 cannonball 光压模型。

    小推力任务质量在线衰减，固定质量的 :class:`SolarRadiationPressure` 会
    在整个传播区间用同一初始质量。本类只存截面积 ``area``（m²）与 ``cr``，
    质量在每个增广状态里取出，因此 ``a = flux·P·(1AU/r)²·cr·area/m`` 随
    推进耗质量自动更新。

    Args:
        area: 航天器迎风截面积，单位 m²。
        cr: 辐射反射系数，默认 1.5。
        shadow: 阴影模型（注入）；``None`` 表示全光照。
    """

    def __init__(
        self,
        area: float,
        cr: float = 1.5,
        shadow: ConicalShadowModel | None = None,
    ) -> None:
        self._area = float(area)
        self._cr = float(cr)
        self._shadow = shadow
        if self._area <= 0:
            raise ValueError("area must be positive")

    @property
    def area(self) -> float:
        return self._area

    @property
    def cr(self) -> float:
        return self._cr

    @property
    def shadow(self) -> ConicalShadowModel | None:
        return self._shadow

    def to_rust_spec(self, system) -> tuple | None:
        """序列化为 ``("srp_variable_mass", area, cr, shadow_bodies)``。"""
        shadow_bodies = list(self._shadow.bodies) if self._shadow is not None else []
        return ("srp_variable_mass", self._area, self._cr, shadow_bodies)
